AuthBridge/AuthProxy/README.md

```diff
@@ -1,113 +1,263 @@
+AuthProxy is a **JWT validation and token exchange proxy** for Kubernetes workloads. It enables secure service-to-service communication by intercepting and validating incoming tokens and transparently exchanging them for tokens with the correct audience for downstream services.
 
-1. **Auth Proxy** (port 8080) - Receives HTTP traffic and checks for "kagenti" authorization
-2. **Demo App** (port 8081) - A demo application that validates "kagenti_new" authorization and responds accordingly
+## What AuthProxy Does
 
-## Deployment
+AuthProxy solves a common challenge in microservices architectures: **how can a service call another service when each service expects tokens with different audiences?**
+
+### The Problem
+
+```
+┌─────────────┐                      ┌─────────────┐
+│   Caller    │ ── Token A ──────────│   Target    │  ❌ REJECTED
+│ (aud: svc-a)│    (aud: authproxy)  │ (expects    │     Wrong audience!
+└─────────────┘                      │  aud: svc-b)│
+                                     └─────────────┘
+```
+
+When a caller obtains a token from an identity provider, the token is scoped to a specific audience. If the caller wants to access a different service that expects a different audience, the request will be rejected.
+
+### The Solution
+
+AuthProxy intercepts outgoing requests, validates the caller's token, and exchanges it for a new token with the correct audience:
+
+```
+┌─────────────┐     ┌──────────────────────────┐     ┌─────────────┐
+│   Caller    │ ──► │       AuthProxy          │ ──► │   Target    │  ✅ AUTHORIZED
+│             │     │  1. Validate token       │     │             │
+│ Token:      │     │  2. Exchange for new aud │     │ Expects:    │
+│ aud=authproxy     │  3. Forward request      │     │ aud=target  │
+└─────────────┘     └──────────────────────────┘     └─────────────┘
+                              │
+                              ▼
+                    ┌─────────────────┐
+                    │    Keycloak     │
+                    │ (Token Exchange)│
+                    └─────────────────┘
+```
+
+## Components
+
+AuthProxy consists of two main components that work together:
+
+### 1. Auth Proxy (`main.go`)
+
+A Go HTTP proxy that:
+- Receives incoming requests on port **8080**
+- Validates JWT tokens using JWKS (JSON Web Key Set)
+- Checks token claims: `issuer`, `audience`, and optionally `scope`
+- Forwards validated requests to the target service
+- Returns `401 Unauthorized` for invalid tokens
+
+**Configuration via environment variables:**
+| Variable | Description | Example |
+|----------|-------------|---------|
+| `JWKS_URL` | URL to fetch public keys for JWT validation | `http://keycloak:8080/realms/demo/.../certs` |
+| `ISSUER` | Expected token issuer | `http://keycloak:8080/realms/demo` |
+| `AUDIENCE` | Expected token audience | `authproxy` |
+| `TARGET_SERVICE_URL` | URL to forward requests to | `http://target-service:8081` |
```

Comment:
It can only forward to one target service? I thought it could handle multiple target services

Reply:
for this demo, just one. Small steps...